<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Document</title>
</head> 
<body>
    <script>
        let userJson = '{"name" : "Petr", "age" : 25, "isAdmin" : true}';
        let user = JSON.parse(userJson); 
        
        user.dateCreated = '2020-08-19';
        user.address = {
            city : "Moscow",
            postalCode : "610001"
        };
        
//        console.log(user);
//        Преобразование объекта в строку JSON
        
//        let user2 = JSON.stringify(user,['name','dateCreated','age','address','city','postalCode'],4);
//        
//        let user3 = JSON.parse(user2,(key,value)=>{
//            if(key == 'dateCreated'){
//                value = new Date(value);
//            }
//            return value;
//        })
//        console.log(user3);
        
        //Асинхронность и синхронность
  /*      
        const async = (a) => {
            setTimeout(()=>{
                const b = a + 1;
                return b;
            },4000);
            return 5000;
        };
        console.log(async(5));*/
      
        //Callback - функция, которую в определенный момент запускает другая функция
        
//        const async = (a, cb) => {
//            setTimeout(()=>{
//                const b = a + 1;
//                cb(b);
//            },4000);
//            return 5000;
//        };
//        console.log(async(5,(data)=>{
//            console.log(data);
//        }));
        
        
//Promice
//Промис - это объект в котором есть свойства resolve и reject и эти свойства
//назначаются  в зависимости от исполнения асинхронности
        
let num = (a) => {
    return new Promise((resolve,reject) =>{
        setTimeout(()=>{
            if(a){
                resolve(a+10);//после resolve вызывается обработчик then
            }
            else{
                reject("Error");//после reject вызывается обработчик catch
            }
        },3000);
        return 5000;
    })
};

num(5)
    .then(data => {
        console.log(data);
    })
    .catch(error => {
        console.log(error);
    })
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
    </script>
</body>
</html>